src/control/server/storage/bdev/firmware.go

```diff
@@ -147,7 +165,58 @@ func getDeviceController(pciAddr string, controllers storage.NvmeControllers) (*
 		}
 	}
 
-	return nil, fmt.Errorf("no NVMe controller found with PCI address %q", pciAddr)
+	return nil, FaultPCIAddrNotFound(pciAddr)
+}
+
+func filterControllers(controllers storage.NvmeControllers, modelID, fwRev string) storage.NvmeControllers {
+	selected := make(storage.NvmeControllers, 0, len(controllers))
+	for _, ctrlr := range controllers {
+		if filterMatches(modelID, ctrlr.Model) &&
+			filterMatches(fwRev, ctrlr.FwRev) {
+			selected = append(selected, ctrlr)
+		}
+	}
+	return selected
+}
+
+func filterMatches(filterStr, actualStr string) bool {
+	return filterStr == "" || strings.ToUpper(actualStr) == strings.ToUpper(filterStr)
+}
+
+// UpdateFirmware updates the NVMe device controller firmware.
+func (p *Provider) UpdateFirmware(req FirmwareUpdateRequest) (*FirmwareUpdateResponse, error) {
+	if p.shouldForward(req) {
+		return p.fwFwd.Update(req)
+	}
+
+	if len(req.FirmwarePath) == 0 {
```

Comment:
Was it decided not to have a default path for the firmware file?

Reply:
Last we discussed it, the firmware packaging plan was still a bit up in the air. Likely we'll be supplying RPMs, at which point there will be a good default location to check. Right now we're not making any assumptions on how they got the file to all the hosts, though.

Reply:
in other areas . I think we are starting to make assumptions based on rpm default locations so we could start to do that now.

Reply:
I think I'll leave that alone for now. There's a lot of intelligence we could add in with the payload manifest files, but I'm inclined to leave any dealings with default values alone until then. The firmware update RPMs will include firmware files for different devices, which we have no way to figure out without the manifest.